urlb54e207112_Ghuillaume_CSA_tgz-pJ8-M0_ApplicationJ8/src/M2/Configuration.java

```diff
@@ -0,0 +1,370 @@
+	}
+	
+	public final Attachment addLink(String name, RoleTo r, PortRequis p) {
+		Attachment a = null;
+		if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			a = new Attachment(name);
+			a.bind(r, p);
+			this.attachments.add(a);
+			this.tags.add(a.getName());
+		}
+		return a;
+	}
+	
+	
+	// Bindings (4)
+	
+	public final Binding addLink(String name, RoleFrom r, PortConfigFourni p) {
+		Binding b = null;
+
+		if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			b = new Binding(name);
+			b.bind(r, p);
+			this.bindings.add(b);
+			this.tags.add(b.getName());
+		}
+		return b;
+	}
+	
+	public final Binding addLink(String name, RoleTo r, PortConfigRequis p) {
+		Binding b = null;
+
+		if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			b = new Binding(name);
+			b.bind(r, p);
+			this.bindings.add(b);
+			this.tags.add(b.getName());
+		}
+		return b;
+		
+	}
+	
+	public final Binding addLink(String name, PortConfigRequis pc, PortRequis p) {
+		Binding b = null;
+		
+		if(pc == null) {
+			Trace.printError("While trying to bind, portConfigRequis is null");
+		}
+		else if(p == null) {
+			Trace.printError("While trying to bind, portRequis is null");
+		}
+		else if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			b = new Binding(name);
+			b.bind(pc, p);
+			this.bindings.add(b);
+			this.tags.add(b.getName());
+		}
+		return b;
+	}
+	
+	public final Binding addLink(String name, PortConfigFourni pc, PortFourni p) {
+		Binding b = null;
+
+		if(this.tags.contains(name)) {
+			Trace.printError("Tag " + name + " is not available, service as not been created");
+		}
+		else {
+			b = new Binding(name);
+			b.bind(pc, p);
+			this.bindings.add(b);
+			this.tags.add(b.getName());
+		}
+		return b;
+	}
+	
+	public final void addLink(Binding b) {
+		if(this.tags.contains(b.getName())) {
+			Trace.printError("Binding already added or name already picked");
+		}
+		else {
+			this.bindings.add(b);
+			this.tags.add(b.getName());
+		}
+	}
+	
+	
+	// Links management
+	
+	public final void callAttachments(Interface sender, String message) {
+
+		ArrayList<Interface> receiver = new ArrayList<Interface>();
+		
+		// Find all receiver in attachments list
+		Iterator<Attachment> itAttach = this.attachments.iterator();
+		while(itAttach.hasNext()) {
+			Interface dest = itAttach.next().getReceiver(sender);
+			if(dest != null) {
+				receiver.add(dest);
+			}
+		}
+		
+		// Activate all receivers
+		Iterator<Interface> itIface = receiver.iterator();
+		while(itIface.hasNext()) {
+			Interface dest = itIface.next();
+			dest.activate(message);
+		}
+	}
+	
+	public final void callBindings(Interface sender, String message) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.